types/parse_blob_txn.go

```diff
@@ -0,0 +1,372 @@
+package types
+
+// Minimal Blob Transaction parser for txpool purposes
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/holiman/uint256"
+	"github.com/protolambda/go-kzg/eth"
+)
+
+const (
+	FieldElementsPerBlob = 4096
+
+	FieldElementSize = 32
+
+	BlobSize = FieldElementsPerBlob * FieldElementSize // blob size in bytes
+
+	ProofSize = 48 // kzg proof size
+
+	MaxBlobsPerBlock = 4
+)
+
+type wrapper struct {
+	proof                    eth.KZGProof
+	sigOffset                int // where the 65 byte signature starts in the payload
+	sigHashStart, sigHashEnd int // the portion of the payload to hash to get the signing hash
+
+	chainID                uint256.Int
+	nonce                  uint64
+	maxPriorityFeePerGas   uint256.Int
+	maxFeePerGas           uint256.Int
+	gas                    uint64
+	creation               bool
+	value                  uint256.Int
+	dataLen                int
+	dataNonZeroLen         int
+	maxFeePerDataGas       uint256.Int
+	accessListAddressCount int
+	accessListKeyCount     int // number of storage keys in all access lists
+
+	blobHashesOffset int
+	numBlobHashes    int
+
+	blobKzgsOffset int
+	numBlobKzgs    int
+
+	blobsOffset int
+	numBlobs    int
+}
+
+type wrapperBlobSequence struct {
+	payload []byte
+	num     int
+}
+type wrapperKzgSequence struct {
+	payload []byte
+	num     int
+}
+type wrapperBlob []byte
+
+func (s wrapperBlobSequence) Len() int {
+	return s.num
+}
+func (s wrapperBlobSequence) At(i int) eth.Blob {
+	if i >= s.num {
+		return nil
+	}
+	r := wrapperBlob(s.payload[i*BlobSize : i*BlobSize+BlobSize])
+	return &r
+
+}
+func (s wrapperKzgSequence) Len() int {
+	return s.num
+}
+func (s wrapperKzgSequence) At(i int) eth.KZGCommitment {
+	kzg := eth.KZGCommitment{}
+	if i >= s.num {
+		return kzg
```

Comment:
should we return a reference for the function? this way you can return nil and make it easier for caller to distinguish return?

Reply:
this is part of the interface expected by the eth. library in go-kzg -- it's idiomatic in geth code at least to always pass small fixed length arrays like these by value.